kotlinx-coroutines-core/common/src/internal/ConcurrentLinkedList.kt

```diff
@@ -203,7 +203,7 @@ internal abstract class Segment<S : Segment<S>>(val id: Long, prev: S?, pointers
      * There are no pointers to this segment from outside, and
      * it is not a physical tail in the linked list of segments.
      */
-    override val removed get() = cleanedAndPointers.value == maxSlots && !isTail
+    override val removed get() = !isTail && cleanedAndPointers.value == maxSlots
 
     // increments the number of pointers if this segment is not logically removed.
     internal fun tryIncPointers() = cleanedAndPointers.addConditionally(1 shl POINTERS_SHIFT) { it != maxSlots || isTail }
```

Comment:
(Duplicating the bug report here for documentation purposes) The changes to `removed` were introduced to fix a bug that caused a segment that was already seen to be logically removed to become alive once again. However, this fix is not sufficient.

Consider the following execution: in a tail segment that has `cleanedAndPointers.value == max_slots`, `tryIncPointers()` is executed concurrently with adding a new segment to the tail's end. First, `tryIncPointers()` observes that this segment is, in fact, the tail. Second, the segment stops being the tail and becomes logically removed. Finally, `tryIncPointers()` finishes, adding a new pointer and raising the segment from the dead.

The suggested fix (which was employed in the formal proof):
```kotlin
override val removed get() = cleanedAndPointers.value == maxSlots
internal fun tryIncPointers() = cleanedAndPointers.addConditionally(1 shl POINTERS_SHIFT) { it != maxSlots }
internal fun decPointers() = cleanedAndPointers.addAndGet(-(1 shl POINTERS_SHIFT)) == maxSlots
fun onSlotCleaned() {
    if (cleanedAndPointers.incrementAndGet() == maxSlots) remove()
}
```
This way, a once-removed segment can't return to life.

Then, in `remove()`, don't fail but instead exit early if it turns out that the segment is the tail one.

Reply:
Thanks for the duplication, I will fix it in a separate PR